### ISSUE TITLE
[info_task_server] race condition in `is_interruptible`

### ISSUE THREAD

**Issue by marc-hanheide** [strands-project/aaf_deployment#312]
So, here's the observation:
Since `15.3.2016, 14:35:05` the robot stood in the charging station and the schedule kept filling up with tasks without executing them:
![image](https://cloud.githubusercontent.com/assets/1153084/13793167/e70d0886-eaf5-11e5-90fd-cfae749f5181.png)
![image](https://cloud.githubusercontent.com/assets/1153084/13793173/eecb5d84-eaf5-11e5-8ed1-0c0eabf98cd9.png)

A call to `rosservice /info_task_server_is_interruptible` reported false. That explains the filling up of the schedule, though @hawesie and @mudrole1 might want to check if the behaviour in this case is intended as a call to get the `/active_task` from the executor also reported an empty task and not the one that was in fact still active.
The culprit in my view lies [here](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L49-L53) and [here](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L73-L79). Assume, the info_terminal_gui is running, but the info_terminal task has been preempted. When a user now clicks anything on the screen (changes the active screen), [`button_pressed_callback()`](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L73-L79) is called and sets `is_interruptible` to `False`. It will never be set to true again in [`execute()`](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L39-L63) as obviously the task is not executing anymore... leaving the whole system dead-locked. I think just resetting `is_interruptible` to True [here](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L54) is definitely requires, but at the same time we mustn't set the flag to False in the first place if a task isn't even running. I would even suggest to only register the [subscriber](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/info_task/scripts/info_task_server.py#L28) when a task is active, and unsubscribe when the task finished. Opinions?

**Comment by marc-hanheide** [comment]
Just to add and check this is intended: If we now call `/task_executor/set_execution_status` to `false` it won't allow it. Which does make sense, but it's also a bit troublesome as one obviously cannot tell the executors to stop executing _after_ a currently not interruptible task has finished... @hawesie

**Comment by cdondrup** [comment]
One thing I don't understand: if the info task server is preempted, how does setting interruptible to false have any influence? If it is preempted it should not be in the schedule and therefore this shouldn't do anything. I will have a look at the code and fix it asap.

**Comment by marc-hanheide** [comment]
That's what I'm also not clear about... That's why I copied @hawesie in.

**Comment by hawesie** [comment]
Does the executor  think the task is still executing? What does the `/task_executor/get_execution_status` service return?

**Comment by hawesie** [comment]
Do you have the task status logs of the offending task? E.g. from the database or `rosrun task_executor task_status.py` (which only prints live)

**Comment by marc-hanheide** [comment]
I tried  `/task_executor/set_execution_status` to false to stop it, but it wouldn't. It stayed True. But `/task_executor/get_active_task` returned an empty task.

you can browse the database by remotely logging into the control PC... see https://github.com/strands-project/strands_management/wiki/Y3-Integration-Care-Scenarios#using-guacamole-for-remote-access and browse what you may be looking for yourself. It started yesterday at around 14:35 (local time, so 13:35 UTC). In the roslog database you should find all executor events.

**Comment by marc-hanheide** [comment]
and it happened again (18/03/16 15:00 CET):

`info_task_server_is_interruptible` reports `False` leading to this:
![image](https://cloud.githubusercontent.com/assets/1153084/13879896/258e89b0-ed1a-11e5-9707-b7d6f6347d39.png)

**Comment by hawesie** [comment]
My two earlier questions are still pertinent:

> Does the executor think the task is still executing? What does the /task_executor/get_execution_status service return?
> 
> Do you have the task status logs of the offending task? E.g. from the database or rosrun task_executor task_status.py (which only prints live)

**Comment by marc-hanheide** [comment]
@hawesie I think I answered it in https://github.com/strands-project/aaf_deployment/issues/312#issuecomment-197552315 ?? What are missing?

**Comment by hawesie** [comment]
The return value of `get_execution_status` not `set_execution_status` when this error is happening, and maybe a dump of the logs? I'm a bit scared to go rooting around on the production system.

**Comment by marc-hanheide** [comment]
Oh, I think I said that... at least I meant it when I said

> It stayed True. 

So, yes, I checked `get_execution_status`, it reported it was executing, but still reported the current_task as empty.

Not sure it's safer when I root through the system, I thought might be eaier for you to search yourself, what you are looking for, but here are `task_event` logs for the tasks that I deem relevant, namely the ones _before_ it got stuck (I think 131 is the one to look at, but I'm not sure) up to the first on the above list (134). It shows all events reported for those tasks.

## Task 131:

```
/* 1 */
{
    "_id" : ObjectId("56ebea37b63c4d58300a868f"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 0,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T11:44:55.624Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 6.2197e+08
    },
    "event" : 1,
    "description" : ""
}

/* 2 */
{
    "_id" : ObjectId("56ebef13b63c4d58300a8ef1"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:05:39.712Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 7.09277e+08
    },
    "event" : 3,
    "description" : ""
}

/* 3 */
{
    "_id" : ObjectId("56ebef15b63c4d58300a8ef2"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:05:41.501Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 4.98232e+08
    },
    "event" : 4,
    "description" : ""
}

/* 4 */
{
    "_id" : ObjectId("56ebef17b63c4d58300a8ef3"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:05:43.509Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 5.07091e+08
    },
    "event" : 5,
    "description" : ""
}

/* 5 */
{
    "_id" : ObjectId("56ebef17b63c4d58300a8ef4"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:05:43.515Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 5.12773e+08
    },
    "event" : 8,
    "description" : ""
}

/* 6 */
{
    "_id" : ObjectId("56ebf05bb63c4d58300a92a4"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:11:07.567Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 5.64245e+08
    },
    "event" : 10,
    "description" : ""
}

/* 7 */
{
    "_id" : ObjectId("56ebf05bb63c4d58300a92a5"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 131,
        "execution_time" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T12:11:07.572Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 5.698e+08
    },
    "event" : 15,
    "description" : ""
}
```

## Task 132:

```
/* 1 */
{
    "_id" : ObjectId("56ebeb63b63c4d58300a88db"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 132,
        "execution_time" : {
            "secs" : 0,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T11:49:55.662Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 6.57483e+08
    },
    "event" : 1,
    "description" : ""
}
```

## Task 133:

```
/* 1 */
{
    "_id" : ObjectId("56ebeb63b63c4d58300a88db"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 132,
        "execution_time" : {
            "secs" : 0,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T11:49:55.662Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 6.57483e+08
    },
    "event" : 1,
    "description" : ""
}

```

## Task 134:

```
/* 1 */
{
    "_id" : ObjectId("56ebeb63b63c4d58300a88db"),
    "task" : {
        "start_node_id" : "ChargingPoint",
        "max_duration" : {
            "secs" : 293,
            "nsecs" : 0
        },
        "task_id" : 132,
        "execution_time" : {
            "secs" : 0,
            "nsecs" : 0
        },
        "end_before" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "end_node_id" : "ChargingPoint",
        "priority" : 1,
        "start_after" : {
            "secs" : 1.4583e+09,
            "nsecs" : 0
        },
        "arguments" : [],
        "action" : "/info_task_server"
    },
    "_meta" : {
        "stored_type" : "strands_executive_msgs/TaskEvent",
        "inserted_by" : "/scheduled_task_executor",
        "stored_class" : "strands_executive_msgs.msg._TaskEvent.TaskEvent",
        "inserted_at" : ISODate("2016-03-18T11:49:55.662Z")
    },
    "time" : {
        "secs" : 1.4583e+09,
        "nsecs" : 6.57483e+08
    },
    "event" : 1,
    "description" : ""
}
```

**Comment by hawesie** [comment]
Do you see "Pausing execution" (https://github.com/strands-project/strands_executive/blob/hydro-release/task_executor/src/task_executor/base_executor.py#L368) in the logs when you try to set the execution status to false. It seems like it should get that far.

**Comment by hawesie** [comment]
Or "Waiting for signal that I can finish after timeout"? (https://github.com/strands-project/strands_executive/blob/hydro-release/task_executor/src/task_executor/sm_base_executor.py#L41)

Which task is actually hanging? Is it 131? The logs indicate that 131 failed execution. I don't really understand why the next task doesn't start if the failure is logged.

Is the general consensus that this is an execution failure when non-interruptible and this is causing the deadlock?

**Comment by marc-hanheide** [comment]
I think, the executor is mostly behaving correctly. The problem is that the task is not interruptible. And it doesn't interrupt it... I'm not certain that 131 is the culprit. 

Look at the full log: https://gist.github.com/anonymous/d78b58b4efce3bcf1f41

**Comment by marc-hanheide** [comment]
This might be different issue, but it's also a problem:
![image](https://cloud.githubusercontent.com/assets/1153084/13885720/1686a7b2-ed34-11e5-92c2-1e3fc4dad93f.png)

So the current schedule is empty. Which is strange as a task was just started:
![image](https://cloud.githubusercontent.com/assets/1153084/13885701/f888af76-ed33-11e5-8d58-3a16c8cbb9ae.png)

and did never finish... This also caused the robot to wonder of at night.

**Comment by marc-hanheide** [comment]
the key issue should now be fixed in #316 . There are still some unclarities about what the executor did here in this situations, I think the scheduled_task_executor should have be tested on non-interruptible tasks and making sure that it handles this correctly in terms of the publishing of the `/current_schedule` and the outcome of `~get_active_task`. But this is another issue that @hawesie might want to put on the issues for `strands_executive`. This issues is believed to be solved now.
